Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,326 @@
+/-
+Copyright (c) 2023 Newell Jensen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+The finite Coxeter groups are classified as the four infinite families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter system.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ) : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relations corresponding to a Coxeter matrix. -/
+@[nolint unusedArguments]
+def ofMatrix [M.IsCoxeter] : B × B → FreeGroup B :=
```

Comment:
Can you remove the assumption `[M.IsCoxeter]` from this definition and the next ones, when it plays no role? Typically, in definitions, try to minimize the assumptions, and only put these in theorems.

Reply:
It was suggested that I add this in a previous review.  See next comment below as well.

Reply:
I said it was important to have the assumption `[M.IsCoxeter]` the definition of `Matrix.CoxeterGroup`, because we do not want to consider `M.CoxeterGroup` for a junk matrix. For the definitions in the section `CoxeterGroup.Relations`, it is more a matter of taste; you may certainly follow @sgouezel suggestion.

Reply:
I am very strongly in favor of your second proposal. There is just something fishy with universes -- there is no reason why `B` and `W` should be in the same universe. In the definition of `IsCoxeterGroup`, I would require `B` to be in `Type u`, as you do, but then I would prove a lemma saying that if you have a Coxeter system in `Type v`, then you can also build one in `Type u`, so that your group indeed satisfies `IsCoxeterGroup`. 

You can see how this is done for the free group in https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean, where the definition `IsFreeGroup` requires a basis in the same universe as the group, but the lemma `isFreeGroup` says that, if you have a basis in any universe, then your group satisfies `IsFreeGroup`.

Reply:
> I am very strongly in favor of your second proposal. There is just something fishy with universes -- there is no reason why `B` and `W` should be in the same universe. In the definition of `IsCoxeterGroup`, I would require `B` to be in `Type u`, as you do, but then I would prove a lemma saying that if you have a Coxeter system in `Type v`, then you can also build one in `Type u`, so that your group indeed satisfies `IsCoxeterGroup`.
> 
> You can see how this is done for the free group in https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean, where the definition `IsFreeGroup` requires a basis in the same universe as the group, but the lemma `isFreeGroup` says that, if you have a basis in any universe, then your group satisfies `IsFreeGroup`.

@sgouezel Okay, so I took a look at the link above and what you mentioned and started playing with trying to get it to work but having some issues following what you did.  I haven't figured out how to make it work with this extra *layer of matrix* that I have.  Nor how I would then show that this matrix must be `IsCoxeter`.  From what it seems I will need the signatures to be:

```
namespace CoxeterSystem

variable {B B' W : Type*} [Group W]

variable {M : Matrix B B ℕ} {M' : Matrix B' B' ℕ}

instance funLike : FunLike (CoxeterSystem M W) (Matrix B B ℕ) (fun _ => W) :=  sorry

protected def reindex (cs : CoxeterSystem M W) (e : Matrix B B ℕ ≃ Matrix B' B' ℕ) : 
        CoxeterSystem M' W := sorry

protected lemma injective (cs : CoxeterSystem M W) : Function.Injective cs := sorry

lemma isCoxeterGroup (cs : CoxeterSystem M W) : IsCoxeterGroup W :=
        ⟨Set.range cs, sorry, ⟨sorry, sorry⟩⟩

end CoxeterSystem
```

does this seem to be headed in the *general* direction?   I am not completely confident these signatures are the way they should be. For the signatures above, I am currently hung up on how to construct the matrix for the existential in `isCoxeterGroup` so if you have any advice on how to do that properly I am keen to hear it.

Reply:
Instead of your `e`, I would use an equiv between `B` and `B'`. With this equiv, you can reindex your matrix `M : Matrix B B ℕ` to get a matrix `Matrix B' B' ℕ`, and this matrix will be Coxeter (it's a copy of the original matrix, after all!)

Reply:
@sgouezel I finally got some time to work on this a bit again after being away.  I just pushed what I currently have but was hoping you could take a quick look and give me your opinion if you think the direction I am going seems okay?  I am mostly following what is done in `IsFreeGroup`.  One of the things I am uncertain of is how to show the map is injective.  When taking a look at how `FreeGroup.of_injective` does it, it seems like I should be able to do something similar since, in theory, `FreeGroup` is a `Quot` after all, but I am unfamiliar with the details still so this may be an ill assumption.  If you know of an easier way to prove this, I am all ears.  In regards to the other proof still left over I think using `QuotienGroup.congr` seems right(?) and still working through the details to show that there should be some map connecting the normal closures of the relations for `B` and `B'`.

Reply:
It seems to me there are some missing lemmas about presented groups: the first thing to prove is that, if you have an equiv `e` between two types alpha and beta, and relations inside alpha, then the group they define is isomorphic to the group defined by these relations pushed by `e` in beta (and moreover the isomorphism, in some sense, lifts `e`). Does that make sense? Then what you need for Coxeter groups should be a particular case of this one.

Reply:
This makes sense and hopefully I can find a proof for this when I get some more time to work on this in a couple weeks to push this PR over the line.  Thanks for the feedback!

Reply:
@sgouezel I haven't been able to figure out how to "lift" the equivalence of the types correctly within the isomorphism.  The below code compiles if you copy and paste it above `end ToGroup` in `PresentedGroup.lean`.

```
@[simp]
def relationsEquiv (e : α ≃ β) : Set (FreeGroup α) ≃ Set (FreeGroup β) :=
  Equiv.Set.congr (FreeGroup.freeGroupCongr e).toEquiv

lemma relationsNormalClosureMap (e : α ≃ β) :
    (Subgroup.normalClosure rels).map (FreeGroup.freeGroupCongr e) =
    (Subgroup.normalClosure (relationsEquiv e rels)) := by
  ext b
  simp only [Subgroup.mem_map, MonoidHom.coe_coe, relationsEquiv]
  constructor
  · rintro ⟨a, ha, hx⟩
    apply Subgroup.subset_normalClosure
    use a
    constructor
    · sorry
    · exact hx
  · sorry

def relationsMulEquiv (e : α ≃ β) : PresentedGroup rels ≃* PresentedGroup (relationsEquiv e rels) :=
  QuotientGroup.congr (Subgroup.normalClosure rels) (Subgroup.normalClosure (relationsEquiv e rels))
    (FreeGroup.freeGroupCongr e) (relationsNormalClosureMap e)
```

`relationsNormalClosureMap` is what is needed but maybe there is a better way of going about this that you could help point me to?  That, or maybe I am on the right track but just need to be more clever with regards to how I am choosing `use a`?  Not really sure what to do at this point.

Reply:
@sgouezel I added `equivCoxeterGroup` to the current branch which is based on #9185 and #9403 

However, I am still stuck at trying to prove `Injective cs`.  With what I currently have, I would need to show that `Injective (PresentedGroup.of (rels := CoxeterGroup.Relations.toSet M)`  I was looking over mathlib and trying to find some inspiration, which I did with the ping pong lemma in `CoprodI.lean` which shows that `Injective (FreeGroup \alpha ->* G)` but it has many assumptions (disjoint etc) so I am unsure if it would be applicable for Coxeter groups.  Other than that, I am wondering if I need to do something similar to `FreeGroup \alpha` in that the word reductions for the Coxeter groups are what would make the hom injective?  

Any suggestions?